### PR TITLE
Allow for additional ansible parameters to be passed in; Exit on errors;

### PR DESCRIPTION
**Pull request by jfwood**

### CODE REVIEW

File: ansibleflow/run.py [hunk 1]
```diff
@@ -18,9 +18,12 @@ def convert_var_filename_to_arg(filename):
     return ' -e @{0}'.format(os.path.abspath(filename))
 
 
-def build_ansible_command(playbook, target, environment):
+def build_ansible_command(playbook, target, environment, arguments):
     command = 'ansible-playbook'
 
+    if arguments.verbosity > 0:
+        command += ' -{0}'.format('v' * arguments.verbosity)
```
Comment (jmvrbanac) [inline ansibleflow/run.py:25]: Shouldn't this be a target configuration option instead of bringing this in via a cli argument?

File: ansibleflow/app.py [hunk 2]
```diff
@@ -50,13 +50,29 @@ def setup_argument_parser():
...
+    run_parser.add_argument(
+            '--env',
+            type=str,
+            default='default')
+    run_parser.add_argument(
+            "-v",
+            "--verbosity",
+            action="count",
+            default=0,
+            help="increase output verbosity")
```
Comment (jmvrbanac) [inline ansibleflow/app.py:68]: This is actually verbosity for ansible, not for ansible-flow. This might be rather confusing.

File: ansibleflow/app.py [hunk 3]
```diff
@@ -50,13 +50,29 @@ def setup_argument_parser():
...
+    run_parser.add_argument(
+            '--file',
+            type=str,
+            default=None)
+    run_parser.add_argument(
+            '--extra-vars',
+            nargs='*')
```
Comment (jmvrbanac) [inline ansibleflow/app.py:75]: Why are extra-vars being passed in via the cli? Shouldn't this just be a target or environment cfg option? The whole purpose of ansible-flow is to pre-configure everything so people can run a target against an environment the same way everytime.

File: ansibleflow/config.py [hunk 4]
```diff
@@ -106,5 +106,6 @@ def get_config(filename='./project.yml'):
     global _config  # Ugly, but it works for now
 
     if not _config:
-        _config = load_config(os.path.abspath(filename))
+        _config = load_config(os.path.abspath(
+                filename or './project.yml'))
```
Comment (jmvrbanac) [inline ansibleflow/config.py:110]: Why the change? `filename` is defaulted to `./project.yml` currently. Am I missing something?

File: ansibleflow/app.py [hunk 5]
```diff
@@ -50,13 +50,29 @@ def setup_argument_parser():
...
+    run_parser.add_argument(
+            "-v",
+            "--verbosity",
+            action="count",
+            default=0,
+            help="increase output verbosity")
+    run_parser.add_argument(
+            '--file',
+            type=str,
+            default=None)
```
Comment (jmvrbanac) [inline ansibleflow/app.py:72]: When I first looked at this I was like "file"? I see now that you're referring to the project config. Perhaps we could name it something more descriptive then "file"?

File: ansibleflow/run.py [hunk 6]
```diff
@@ -42,12 +45,20 @@ def build_ansible_command(playbook, target, environment):
...
+
     return command
 
 
 def run(target_name, env_name, arguments, dry_run=False):
-    target = get_config().targets.get(target_name, None)
-    environment = get_config().environments.get(env_name, None)
+    target = get_config(filename=arguments.file).targets.get(target_name)
+    environment = get_config(filename=arguments.file).environments.get(
+            env_name)
```
Comment (jmvrbanac) [inline ansibleflow/run.py:61]: This is a rather ugly line wrap. It would probably be better to just assign the config to a local var so that you don't have to use the line wrap at all.

File: ansibleflow/venv.py [hunk 7]
```diff
@@ -38,7 +39,9 @@ def execute_under_env(command, os_env=None):
         proc.stdin.write(long_cmd.encode('utf-8'))
         proc.stdin.flush()
 
-        proc.wait()
+        returncode = proc.wait()
+        if returncode:
+            sys.exit(returncode)
```
Comment (jmvrbanac) [inline ansibleflow/venv.py:44]: Am I understanding this correcting, you're only interested in sys.exiting when the returncode is > 0? If you are, It might be good to be explicit with that.

File: ansibleflow/run.py (hunk 1 above)
Comment (jfwood) [inline ansibleflow/run.py:25]: This is a convenience...it's easier to just apply debug to the executed tasks than to locate the target in a yml file and then re-run the target.

File: ansibleflow/app.py (hunk 2 above)
Comment (jfwood) [inline ansibleflow/app.py:68]: I couldn't figure out how to get the remaining text not parsed by ansible-flow to be just available to pass onto ansible-playbook. I would prefer that but argparse seems to want to parse - and -- options anyway even if you use the '*' or 'REMAINING' options for nargs. :\

File: ansibleflow/app.py (hunk 3 above)
Comment (jfwood) [inline ansibleflow/app.py:75]: What is prompting the extra-vars part of this anyway, is a desire to pass a version stamp into the ansible playbooks (for use when downloading docker images for example) from Jenkins, when the ansible-flow call is being made, rather than having to sed-modify the project.yml file to add the extra args.

File: ansibleflow/config.py (hunk 4 above)
Comment (jfwood) [inline ansibleflow/config.py:110]: Yeah I need to retract that...I couldn't get my IDE to read relative paths, so I provided a fully resolved one for testing this way.

File: ansibleflow/app.py (hunk 5 above)
Comment (jfwood) [inline ansibleflow/app.py:72]: Please see note above.

File: ansibleflow/run.py (hunk 6 above)
Comment (jfwood) [inline ansibleflow/run.py:61]: Good point.

File: ansibleflow/venv.py (hunk 7 above)
Comment (jfwood) [inline ansibleflow/venv.py:44]: Yeah basically any non-zero response would be good to terminate processing on, to halt Jenkins from proceeding to the next job in the chain for example.

**Comment by jfwood** [comment]
Thanks for the review John!

File: ansibleflow/app.py (hunk 2 above)
Comment (jmvrbanac) [inline ansibleflow/app.py:68]: So, the way that I've dealt with this before was to just to use the ansible-options value on a given target. This allows for you to specify any number of custom options you want to be passed on to ansible.

File: ansibleflow/app.py (hunk 3 above)
Comment (jmvrbanac) [inline ansibleflow/app.py:75]: That makes sense.

**Comment by jmvrbanac** [comment]
@jfwood Is this PR still relevant? It's been sitting here for quite a while.

**Comment by jfwood** [comment]
Hey John, I've moved off to the Nova team and I think the Keep team is using make files now. Therefore closing this PR out as OBE.
